File: code/matlab/read_example.m
function Data = read_example (file)
%READ_EXAMPLE An example for reading a L3Pilot Common Data Format File
%   This funciton simply iterates over all available datasets and reads
%   them to the output struct. In this case, top-level functions can be
%   used.
%
%   Example usage:
%   - Data = read_example('hdf5_filled_example_matlab_v02.h5');
%
% Copyright Institut fuer Kraftfahrzeuge, Johannes Hiller, 2018
%
%SEE ALSO h5read, h5readatt, h5info

% Get all the infos about the HDF5 file
file_infos = h5info(file);
% Iterate over all datasets and read them
for jj = 1:length(file_infos.Datasets)
    disp(['Found dataset ' file_infos.Datasets(jj).Name ' with a size of '...
        num2str(file_infos.Datasets(jj).Datatype.Size) ' bytes per timestamp and ' ...
        num2str(length(file_infos.Datasets(jj).Datatype.Type.Member)) ...
        ' members'])
    Data.(file_infos.Datasets(jj).Name) = ...
        h5read(file, ['/' file_infos.Datasets(jj).Name]);
end

% Iterate over all datasets in groups
for kk = 1:length(file_infos.Groups)
    disp(['Found group ' file_infos.Groups(kk).Name]);

    
    
    for jj = 1:length(file_infos.Groups(kk).Datasets)
        disp(['Found dataset ' file_infos.Groups(kk).Datasets(jj).Name ' with a size of '...
            num2str(file_infos.Groups(kk).Datasets(jj).Datatype.Size) ' bytes per timestamp and ' ...
            num2str(length(file_infos.Groups(kk).Datasets(jj).Datatype.Type.Member)) ...
            ' members']);
        Data.(file_infos.Groups(1).Name(2:length(file_infos.Groups(1).Name))).(file_infos.Groups(kk).Datasets(jj).Name) = ...
            h5read(file, [file_infos.Groups(kk).Name '/' file_infos.Groups(1).Datasets(jj).Name]);
    end
        
    

end


% Iterate over the top-level attributes and read them
for kk = 1:length(file_infos.Attributes)
    Data.metadata.(file_infos.Attributes(kk).Name) = ...
        h5readatt(file, '/', file_infos.Attributes(kk).Name);
end
end

File: code/python/read_example.py
#!/usr/bin/env python

""" Reads the example file to a dict

    This file shows, how to read the data into a dict, that can be used as any other Python dict.
    Also it shows how to read the attributes available in the file and in a dataset.
 """

from structure_definitions import *

__author__ = "Johannes Hiller"
__version__ = "0.8"
__maintainer__ = "Johannes Hiller"
__email__ = "johannes.hiller@ika.rwth-aachen.de"
__status__ = "development"

# The file to read
h5_file = "l3pilot_cdf_filled_example.h5"
# Initialize the dict for storing the data
h5_dict = dict()
# Open the file
with h5py.File(h5_file, "r") as file:
    # Initialize the ego part of the dict
    h5_dict[str_ego] = dict()
    for item in [list1 for list1 in ego_struct if list1.__class__.__name__ == "EgoTop"]:
        h5_dict[str_ego][item.name] = file[str_ego][item.name]
    # Initialize the lane part of the dict
    h5_dict[str_lan] = dict()
    for item in [list1 for list1 in lane_struct if list1.__class__.__name__ == "LaneTop"]:
        h5_dict[str_lan][item.name] = file[str_lan][item.name]
    # Initialize the dict for the different lanes
    h5_dict[str_lan][str_lan_obj] = dict()
    for item in [list1 for list1 in lane_struct if list1.__class__.__name__ == "LaneSObject"]:
        h5_dict[str_lan][str_lan_obj][item.name] = file[str_lan][str_lan_obj][item.name]
    # Initialize the object part of the dict
    h5_dict[str_obj] = dict()
    for item in [list1 for list1 in object_struct if list1.__class__.__name__ == "ObjectTop"]:
        h5_dict[str_obj][item.name] = file[str_obj][item.name]
    # Initialize the dict for the different objects
    h5_dict[str_obj][str_obj_obj] = dict()
    for item in [list1 for list1 in object_struct if list1.__class__.__name__ == "ObjectSObject"]:
        h5_dict[str_obj][str_obj_obj][item.name] = file[str_obj][str_obj_obj][item.name]
    # Initialize the positioning part of the dict
    h5_dict[str_pos] = dict()
    for item in [list1 for list1 in pos_struct if list1.__class__.__name__ == "PosTop"]:
        h5_dict[str_pos][item.name] = file[str_pos][item.name]

    # Print some information about the file
    print("The file is in version {} of the L3Pilot Common Data Format and was created by {}, {} at {}"
          .format(float(file.attrs["metaData"]["General"]["FormatVersion"]), file.attrs["author"],
                  file.attrs["institution"], file.attrs["creation_date"]))
    # Print some information on a signal in the egoVehicle dataset
    print("The description of the UTCTime is : \"{}\" and its unit is \"{}\""
          .format(file["egoVehicle"].attrs["UTCTime"][0][1], file["egoVehicle"].attrs["UTCTime"][1][1]))

# Print the number of timestamps contained in the file
print("The loaded h5 file contains {} timestamps".format(len(h5_dict["egoVehicle"]["UTCTime"])))

